config/runtime.exs

```diff
@@ -239,3 +240,27 @@ if config_env() == :prod do
       redirect_uri: google_redirect_uri
   end
 end
+
+# OIDC Auth
+auth_oidc_env = System.get_env("AUTH_OIDC")
+
+if auth_oidc_env do
+  auth_oidc =
+    Jason.decode!(auth_oidc_env)
+    # Convert Map to something openid_connect expects, atomic keyed configs
+    # eg. %{"provider" => [client_id: "CLIENT_ID" ...]}
+    |> Enum.reduce(%{}, fn {provider, settings}, acc ->
+      Map.put(
+        acc,
+        provider,
+        FzCommon.FzMap.map_to_keyword_list(
+          settings
+          # Update redirect/callback url to use the external_url
+          |> Map.put("redirect_uri", "#{external_url}/auth/oidc/#{provider}/callback/")
+        )
+      )
+    end)
```

Comment:
I think we can simplify this to a simple pipeline and remove the custom map to keyword list function. It's a rather dangerous function and we don't want people to use it too freely. Plus, as the structure of the settings is flat (correct me if I'm wrong) I don't think we need to recursively parse it anymore.

```elixir
|> Map.new(fn {provider, settings} ->
  {provider,
   settings
   |> Map.take(
     ~w(discovery_document_uri client_id client_secret redirect_uri response_type scope label)
   )
   |> Map.put("redirect_uri", "#{external_url}/auth/oidc/#{provider}/callback/")
   |> Enum.map(fn {k, v} -> {String.to_existing_atom(k), v} end)}
end)
```

Reply:
Thanks @princemaple , this is much cleaner, and yes, the structure of the map is flat so we can get rid of the custom helper. I'll make the changes tomorrow morning and do some quick testing.

Reply:
So openid_connect does create these atoms, but, I believe, not until the Worker is created with the passed in configuration. So existing_atom fails in this instance. I've changed it to `to_atom`, which should be safe because of the previous pipeline `take` which limits the number of potential atoms. @princemaple thoughts?

Reply:
@mdp I know it's a little more verbose, but how about more explicitly mapping the atoms in a Keyword list like this?

```elixir
Map.new(fn {provider, settings} ->
  {provider, [
    discovery_document_uri: settings["discovery_document_uri"],
    client_id: settings["client_id"],
    client_secret: settings["client_secret"],
    redirect_uri: "#{external_url}/auth/oidc/#{provider}/callback/",
    response_type: settings["response_type"],
    scope: settings["scope"],
    label: settings["label"]
  ]
end)
```

Reply:
> So openid_connect does create these atoms, but, I believe, not until the Worker is created with the passed in configuration. So existing_atom fails in this instance.

Not until the worker is started, but you are right that `runtime.exs` is loaded before the libs are loaded and started. I'm happy with the `to_atom` here as it's behind a `take` and load time only. The verbose way also looks fine. I'll leave the decision to you guys.

Reply:
@jamilbk Yeah, I like this better, it's clearer about exactly what's happening.